### PR TITLE
act syntax description

### PR DESCRIPTION
**Pull request by MrChico**
This aims to be a start at addressing #29

### CODE REVIEW

File: docs/act.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+```
+
+A constructor section consists of the following fields:
+
+### interface
+
+Specifying the arguments the constructor takes.
+Example:
+```act
+interface constructor(uint _owner)
```
Comment (xwvvvvwx) [inline docs/act.md:42]:
```suggestion
interface constructor(address _owner)
```

File: docs/act.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+### creates
+Defines the storage layout of the contract.
+
+Example:
+```act
+
+creates
+
+  uint256 totalSupply := _totalSupply
+  mapping(address => uint) balanceOf :=  [CALLER := _totalSupply]
```
Comment (xwvvvvwx) [inline docs/act.md:68]: would be nice to have a little info on the allowed types here, as well as the initializer list syntax.

File: docs/act.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+and consists of the following fields:
+
+
+### interface
+
+Specifying which method the behaviour refers to.
+Example:
+```act
+interface transfer(address to, uint value)
```
Comment (xwvvvvwx) [inline docs/act.md:123]: we should note the allowed types here

File: docs/act.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+Example:
+```act
+iff
+
+  balanceOf[CALLER] >= value
+  CALLVALUE == 0
+```
+
+### cases and state updates
```
Comment (xwvvvvwx) [inline docs/act.md:140]: Maybe worth adding a note that either `returns` or `storage` are required, but not both?

**Review by xwvvvvwx — APPROVED** [review]
nice!!!

Can we add a link to here in the readme somewhere?

File: docs/act.md [hunk 5]
```diff
@@ -11,18 +11,51 @@ Each `behaviour` specification determines how a contract method updates
...
 
-This document is a high level description of the syntax of 
+This document is a high level description of the syntax of
 act specifications.
 For a more formal treatment of the syntax, study the
 definitions of [RefinedAst.hs](https://github.com/ethereum/act/blob/master/src/RefinedAst.hs).
 
+## Types
+
+The types of Act are made up from three basic primitives:
```
Comment (leonardoalt) [inline docs/act.md:28]:
```suggestion
The types of Act consist of three basic primitives:
```

File: docs/act.md [hunk 6]
```diff
@@ -11,18 +11,51 @@ Each `behaviour` specification determines how a contract method updates
...
+Integers, Booleans and ByteStrings. Integers are unbounded,
+with true integer operations. However, as our integer expressions
+will often represent words in the EVM, we allow ourselves a slight
+abuse of notation and denote by `uintN` integers together with the
+constraint that the value fits into a `uintN`.
```
Comment (leonardoalt) [inline docs/act.md:33]: Is this also valid for `int`?

File: docs/act.md [hunk 7]
```diff
@@ -11,18 +11,51 @@ Each `behaviour` specification determines how a contract method updates
...
+
+Using conventional abi types for typing also allows us to specify
```
Comment (leonardoalt) [inline docs/act.md:35]:
```suggestion
Using conventional ABI types for typing also allows us to specify
```

File: docs/act.md [hunk 8]
```diff
@@ -11,18 +11,51 @@ Each `behaviour` specification determines how a contract method updates
...
+
+   x + y
+
+returns x + y
+```
+
+In more verbose terms, this specification would read:
+
+Given any pair of integers x and y, s.t. 0 <= x < 2^256 and
+0 <= y < 2^256, an abiencoded call to the contract `SafeMath`
```
Comment (leonardoalt) [inline docs/act.md:54]:
```suggestion
0 <= y < 2^256, an ABI encoded call to the contract `SafeMath`
```

File: docs/act.md (hunk 6 above)
Comment (MrChico) [reply docs/act.md:33]: yes. The semantics then is that `x : intN` is `x : int s.t. 2 ^ (N - 1) <= x < 2 ^ (N -1 )`

Comment (MrChico) [reply docs/act.md:33]: will add to clarify

File: docs/act.md [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+definitions of [RefinedAst.hs](https://github.com/ethereum/act/blob/master/src/RefinedAst.hs).
+
+## Types
+
+The types of Act consist of three basic primitives:
+Integers, Booleans and ByteStrings. Integers are unbounded,
+with true integer operations. However, as our integer expressions
+will often represent words in the EVM, we allow ourselves a slight
+abuse of notation and denote by `uintN` integers together with the
+constraint that the value fits into a `uintN`.
```
Comment (MrChico) [inline docs/act.md:33]:
```suggestion
will often represent words in the EVM, we allow ourselves a slight
abuse of notation and denote by `uintN/intN` integers together with the
constraint that the value fits into a `uintN/intN`.
```

**Pull request merged**